File: services/chess_service/src/main/java/net/chess_platform/chess_service/chess/move/AbstractMove.java
package net.chess_platform.chess_service.chess.move;

import java.time.Instant;

import net.chess_platform.chess_service.chess.Chessboard;
import net.chess_platform.chess_service.chess.piece.AbstractPiece;
import net.chess_platform.chess_service.chess.piece.Piece;
import net.chess_platform.chess_service.chess.piece.Piece.Color;

public abstract class AbstractMove implements Move {

    private final Position from;

    private final Position to;

    private final Piece.Type piece;

    private final Color color;

    private final Type type;

    private Piece movedPieceInstance;

    private CheckStatus checkStatus;

    private long timestamp;

    public AbstractMove(Piece.Type piece, Color color, Position from, Position to, Type type) {
        this.piece = piece;
        this.color = color;
        this.from = from;
        this.to = to;
        this.type = type;
    }

    @Override
    public boolean validate(Chessboard board) {
        execute(board);
        boolean isKingInCheck = board.isKingInCheck(movedPieceInstance.getColor());
        undo(board);
        return !isKingInCheck;
    }

    @Override
    public void execute(Chessboard board) {
        timestamp = Instant.now().toEpochMilli();
        movedPieceInstance = board.getPiece(from);
        movedPieceInstance.incrementMoveCount();
    }

    @Override
    public void undo(Chessboard board) {
        movedPieceInstance.decrementMoveCount();
        movedPieceInstance = null;
        timestamp = 0;
    }

    @Override
    public Position getFrom() {
        return from;
    }

    @Override
    public Position getTo() {
        return to;
    }

    @Override
    public CheckStatus getCheckStatus() {
        return checkStatus;
    }

    @Override
    public void setCheckStatus(CheckStatus checkStatus) {
        if (checkStatus == null) {
            return;
        }

        this.checkStatus = checkStatus;
    }

    @Override
    public Color getColor() {
        return color;
    }

    @Override
    public AbstractPiece.Type getPiece() {
        return piece;
    }

    @Override
    public long getTimestamp() {
        return timestamp;
    }

    @Override
    public Type getType() {
        return type;
    }

    @Override
    public boolean isPromotion() {
        var to = getTo();
        return getPiece() == Piece.Type.PAWN && (to.row() == 7 || to.row() == 0); 
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public Piece getMovedPieceInstance() {
        return movedPieceInstance;
    }

}

File: app/src/chess/moves/AbstractMove.ts
import { Chessboard } from "../Chessboard";
import { Color } from "../pieces/AbstractPiece";
import { Piece, PieceType } from "../pieces/Piece";
import { CheckStatus, Move, MoveType, Position } from "./Move";

export abstract class AbstractMove implements Move {

    private from: Position;

    private to: Position;

    private piece: PieceType | null;

    private color: Color;

    private type: MoveType;

    private checkStatus?: CheckStatus;

    private timestamp?: number;

    private movedPiece?: Piece;

    constructor(piece: PieceType | null, color: Color, from: Position, to: Position, type: MoveType, checkStatus?: CheckStatus, timestamp?: number) {
        this.piece = piece;
        this.color = color;
        this.from = from;
        this.to = to;
        this.type = type;
        this.checkStatus = checkStatus;
        this.timestamp = timestamp;
    }

    toAlgebraicNotation(): string {
        const notation = this.toAlgebraicNotationImpl();
        if (this.checkStatus === "CHECK") {
            return notation + "+";
        }

        if (this.checkStatus === "CHECKMATE") {
            return notation + "++";
        }

        return notation;
    }

    abstract toAlgebraicNotationImpl(): string;

    validate(board: Chessboard): boolean {
        this.execute(board);
        const isKingInCheck = board.isKingInCheck(this.getColor());
        this.undo(board);
        return !isKingInCheck;
    }

    getPiece(): PieceType | null {
        return this.piece;
    }

    getColor(): Color {
        return this.color;
    }

    execute(board: Chessboard): void {
        const { from } = this;
        this.movedPiece = board.getPiece(from.row, from.col)!;
        this.movedPiece.incrementMoveCount();
    }

    undo(board: Chessboard): void {
        this.movedPiece!.decrementMoveCount();
    }

    getFrom(): Position {
        return this.from;
    }

    getTo(): Position {
        return this.to;
    }

    getCheckStatus(): CheckStatus | undefined {
        return this.checkStatus;
    }

    getTimestamp(): number | undefined {
        return this.timestamp;
    }

    getType(): MoveType {
        return this.type;
    }

    getMovedPieceInstance(): Piece | undefined {
        return this.movedPiece;
    }

}